src/libstrongswan/plugins/pkcs11/pkcs11_private_key.c

```diff
@@ -204,11 +213,140 @@ CK_MECHANISM_PTR pkcs11_signature_scheme_to_mech(pkcs11_library_t *p11,
+							mechanism->mechanism = CKM_SHA384_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA512:
+							mechanism->mechanism =  CKM_SHA512_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA224:
+							mechanism->mechanism =  CKM_SHA224_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA1:
+							mechanism->mechanism =  CKM_SHA1_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA3_256:
+							mechanism->mechanism =  CKM_SHA3_256_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA3_384:
+							mechanism->mechanism =  CKM_SHA3_384_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA3_512:
+							mechanism->mechanism =  CKM_SHA3_512_RSA_PKCS_PSS;
+							break;
+						case HASH_SHA3_224:
+							mechanism->mechanism =  CKM_SHA3_224_RSA_PKCS_PSS;
+							break;
+						default:
+							free(rsa_pkcs_pss_params);
+							free(mechanism);
+							return NULL;
+					}
+					if (!is_mechanism_supported(p11, slot, mechanism))
+					{
+						/* revert to external hashing */
+						mechanism->mechanism =  CKM_RSA_PKCS_PSS;
+					}
+				}
+				switch (rsa_pss_params->hash)
+				{
+					case HASH_SHA256:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA256;
+						break;
+					case HASH_SHA384:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA384;
+						break;
+					case HASH_SHA512:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA512;
+						break;
+					case HASH_SHA224:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA224;
+						break;
+					case HASH_SHA1:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA_1;
+						break;
+					case HASH_SHA3_256:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA3_256;
+						break;
+					case HASH_SHA3_384:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA3_384;
+						break;
+					case HASH_SHA3_512:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA3_512;
+						break;
+					case HASH_SHA3_224:
+						rsa_pkcs_pss_params->hashAlg = CKM_SHA3_224;
+						break;
+					default:
+						free(rsa_pkcs_pss_params);
+						free(mechanism);
+						return NULL;
+				}
+				switch (rsa_pss_params->mgf1_hash)
+				{
+					case HASH_SHA256:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA256;
+						break;
+					case HASH_SHA384:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA384;
+						break;
+					case HASH_SHA512:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA512;
+						break;
+					case HASH_SHA224:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA224;
+						break;
+					case HASH_SHA1:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA1;
+						break;
+					case HASH_SHA3_256:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA3_256;
+						break;
+					case HASH_SHA3_384:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA3_384;
+						break;
+					case HASH_SHA3_512:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA3_512;
+						break;
+					case HASH_SHA3_224:
+						rsa_pkcs_pss_params->mgf = CKG_MGF1_SHA3_224;
+						break;
+					default:
+						free(rsa_pkcs_pss_params);
+						free(mechanism);
+						return NULL;
+				}
+				hash_alg = rsa_pss_params->hash;
```

Comment:
I think this should only be assigned if `mechanism->mechanism` is `CKM_RSA_PKCS_PSS`, not when the mechanism is a specific one that includes the hash. Otherwise, the token would get a hash to hash again and wouldn't produce the correct result. `hash_alg` has to be initialized to `HASH_UNKNOWN` for this to work.